Development/nmos/ocsp_behaviour.cpp

```diff
@@ -0,0 +1,366 @@
+    void details::ocsp_behaviour_thread(nmos::model& model, nmos::experimental::ocsp_state& ocsp_state, load_ca_certificates_handler load_ca_certificates, load_server_certificates_handler load_server_certificates, slog::base_gate& gate)
+    {
+        enum
+        {
+            initial_ocsp_operation,
+            ocsp_behaviour
+        } mode = initial_ocsp_operation;
+
+        std::vector<web::uri> ocsp_uris;
+
+        nmos::details::seed_generator backoff_seeder;
+        std::default_random_engine backoff_engine(backoff_seeder);
+        double backoff = 0;
+
+        ocsp_shared_state state{ load_ca_certificates };
+
+        // continue until the server is being shut down
+        for (;;)
+        {
+            if (with_read_lock(model.mutex, [&] { return model.shutdown; })) break;
+
+            switch (mode)
+            {
+                case initial_ocsp_operation:
+                {
+                    // note: use the same discovery backoff settings for OCSP server retry
+                    if (0 != backoff)
+                    {
+                        auto lock = model.read_lock();
+                        const auto random_backoff = std::uniform_real_distribution<>(0, backoff)(backoff_engine);
+                        slog::log<slog::severities::more_info>(gate, SLOG_FLF) << "Waiting to retry OCSP server for about " << std::fixed << std::setprecision(3) << random_backoff << " seconds (current backoff limit: " << backoff << " seconds)";
+                        model.wait_for(lock, std::chrono::milliseconds(std::chrono::milliseconds::rep(1000 * random_backoff)), [&] { return model.shutdown; });
+                        if (model.shutdown) break;
+                    }
+
+                    // get the list of server certificates chain
+                    const auto server_certificates = load_server_certificates();
+                    const auto server_certificate_chains = boost::copy_range<std::vector<utility::string_t>>(server_certificates | boost::adaptors::transformed([](const nmos::certificate& certificate) { return certificate.certificate_chain; }));
+
+                    try
+                    {
+                        // get OCSP URIs
+                        ocsp_uris = details::get_ocsp_uris(server_certificate_chains, gate);
+                        if (ocsp_uris.size())
+                        {
+                            mode = ocsp_behaviour;
+
+                            // extract the shortest half certificate expiry time from all server certificates
+                            state.next_request = details::half_certificate_expiry_from_now(server_certificate_chains, gate);
+
+                            // construct an OCSP request with the server certificates
+                            state.ocsp_request = details::make_ocsp_request(server_certificate_chains, gate);
+                            if (state.ocsp_request.size())
```

Comment:
```suggestion
                            if (!state.ocsp_request.empty())
```

P.S. @lo-simon, what does it mean if it's empty? If that's unexpected, should we log something?

Reply:
I think we can actually remove the 

line 97
`if (ocsp_uris.size())`
and line106
`if (state.ocsp_request.size())`
as `get_ocsp_uris` and `make_ocsp_request` have already throw for empty
which should be logged in the `try...catch` block.

And that will make the code clear too.

Good point what should we do if there is no OCSP server found in the certificates, then there is no point to have the `ocsp_behaviour_thread` running unless certificates can be changed dynamically.